Multiple elaborations about template literal semantics (#18976)
mdn/content: files/en-us/web/javascript/reference/template_literals/index.md

### String interpolation

Without template literals, when you want to combine output from expressions with strings, you'd [concatenate them](/en-US/docs/Learn/JavaScript/First_steps/Strings#concatenation_using_) using the [addition operator](/en-US/docs/Web/JavaScript/Reference/Operators/Addition) `+`:

```js
const a = 5;
const b = 10;
console.log('Fifteen is ' + (a + b) + ' and\nnot ' + (2 * a + b) + '.');
// "Fifteen is 15 and
// not 20."
```

That can be hard to read – especially when you have multiple expressions.

With template literals, you can avoid the concatenation operator — and improve the readability of your code — by using placeholders of the form `${expression}` to perform substitutions for embedded expressions:

```js
const a = 5;
const b = 10;
console.log(`Fifteen is ${a + b} and
not ${2 * a + b}.`);
// "Fifteen is 15 and
// not 20."
```

> Note that there's a mild difference between the two syntaxes. Addition would coerce the expression to a _primitive_, which calls [`valueOf()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf) in priority; on the other hand, template literal would coerce the expression to a _string_, which calls [`toString()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString) in priority. If the expression has a [`@@toPrimitive`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive) method, string concatenation calls it with `"default"` as hint, while template literal uses `"string"`. This is important for objects that have different string and primitive representations — such as [Temporal](https://github.com/tc39/proposal-temporal), whose `valueOf()` method throws.

Reviewer: ```suggestion
Note that there's a mild difference between the two syntaxes. Addition would coerce the expression to a _primitive_, which calls [`valueOf()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf) in priority; on the other hand, template literal would coerce the expression to a _string_, which calls [`toString()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString) in priority. If the expression has a [`@@toPrimitive`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive) method, string concatenation calls it with `"default"` as hint, while template literal use `"string"`. This is important for objects that have different string and primitive representations — such as [Temporal](https://github.com/tc39/proposal-temporal), whose `valueOf()` method throws.
```

Author: "template literal uses" or "template literals use"? Because "string concatenation" is singular as well.

Reviewer: > "template literal uses" or "template literals use"? Because "string concatenation" is singular as well.

oops yeah — just now updated it. I think the commit suggestion is right now?

Reviewer: …or else “a template literal uses” (if you want it singular)

Author: Yep—this looks good.

Reviewer: …or else “template literal _syntax_”

```js
const t = Temporal.Now.instant();
"" + t; // Throws TypeError
`${t}`; // '2022-07-31T04:48:56.113918308Z'

> ```

Author: Don't know what we want to do about this example. I certainly want to "prepare for the future" and avoid using a contrived example just because there are no builtins with this behavior currently.

Reviewer: > Don't know what we want to do about this example. I certainly want to "prepare for the future" and avoid using a contrived example just because there are no builtins with this behavior currently.

I don’t know what to do about it either — but I think it’s fine to just leave this way for this PR

// "H,i,\\,n,5,!"
```

`String.raw` functions like an "identity" tag if the literal doesn't contain any escape sequences. In case you want an actual identity tag that always works as if the literal is untagged, you can make a custom function that passes the "cooked" (i.e. escape sequences are processed) literal array to `String.raw`, pretending they are raw strings.

```js
const identity = (strings, ...values) => String.raw({ raw: strings }, ...values);
console.log(identity`Hi\n${2 + 3}!`);
// Hi
// 5!
```

> This is useful for many toolings which give special treatments to literals tagged by a particular name.

Reviewer: ```suggestion
This is useful for many tools which give special treatment to literals tagged by a particular name.
```

// { cooked: undefined, raw: "\\unicode" }
```

> Note that the escape sequence restriction is only dropped from _tagged_ templates—not from _untagged_ template literals:

Reviewer: ```suggestion
Note that the escape-sequence restriction is only dropped from _tagged_ templates—not from _untagged_ template literals:
```
